src/MatrixTwitter.js

```diff
@@ -0,0 +1,1040 @@
+        return intent.setAvatarUrl(uri);
+      }).catch(err => {
+        log.error(
+            'Twitter',
+            "Couldn't set new avatar for @%s because of %s",
+            profile.screen_name,
+            err
+          );
+      });
+    }
+    this.storage.cache_user_profile(profile.id, profile.screen_name, profile, ts);
+  });
+}
+
+MatrixTwitter.prototype._create_twitter_client = function (creds) {
+  var ts = new Date().getTime();
+  var client = new Twitter({
+    consumer_key: this.app_auth.consumer_key,
+    consumer_secret: this.app_auth.consumer_secret,
+    access_token_key: creds.access_token,
+    access_token_secret: creds.access_token_secret
+  });
+  client.last_auth = ts;
+  return client;
+}
+
+MatrixTwitter.prototype._get_twitter_client = function (sender) {
+  //Check if we have the account in the cache
+  return this.storage.get_twitter_account(sender).then((creds) => {
+    return new Promise( (resolve, reject) => {
+      if(creds == null) {
+        reject("No twitter account linked.");
+        return;
+      }
+
+      var ts = new Date().getTime();
+      var id = creds.user_id;
+      var client;
+      if(this.tclients.has(id)) {
+        client = this.tclients[id];
+        if(ts - client.last_auth < TWITTER_CLIENT_INTERVAL_MS) {
+          resolve(client);
+          return;
+        }
+
+        log.info("Twitter", "Credentials for %s need to be reevaluated.", sender);
+        client.get("account/verify_credentials", (error, profile) => {
+          if(error) {
+            log.info("Twitter", "Credentials for " + id + " are no longer valid.");
+            log.error("Twitter", error);
+            delete this.tclients[id];//Invalidate it
+            resolve(this._get_twitter_client(sender));
+          }
+          client.profile = profile;
+          this._update_user_timeline_profile(profile);
+          resolve(client);
+        });
+      }
+      else {
+        client = this._create_twitter_client(creds);
+        this.tclients[id] = client;
+        client.get("account/verify_credentials", (error, profile) => {
+          if(error) {
+            delete this.tclients[id];//Invalidate it
+            log.error(
+              "Twitter",
+              "We couldn't authenticate with the supplied access token for %s. Look into this.",
+              id
+            );
+            reject("Twitter account could not be authenticated.");
+            return;
+            //TODO: Possibly find a way to get another key.
+          }
+          client.profile = profile;
+          client.last_auth = ts;
+          this._update_user_timeline_profile(profile);
+          resolve(client);
+        });
+      }
+    });
+  });
+}
+
+MatrixTwitter.prototype._get_user = function (data) {
+  return new Promise((resolve, reject) => {
+    this.app_twitter.get('users/show', data, (error, user) => {
+      if (error) {
+        error = error[0];
```

Comment:
From IRL: The `twitter` library sucks and can sometimes return an array of errors rather than be consistent. We should probably check it is an array to be sure.